helios-client/src/main/java/com/spotify/helios/auth/CrtAuthProvider.java

```diff
@@ -0,0 +1,159 @@
+
+  private final RequestDispatcher dispatcher;
+  private final String username;
+  private final CrtAuthClient crtClient;
+  private final AtomicReference<String> tokenRef;
+
+  public CrtAuthProvider(final RequestDispatcher dispatcher,
+                         final CrtAuthClient crtClient,
+                         final String username) {
+    this.dispatcher = dispatcher;
+    this.crtClient = crtClient;
+    this.username = username;
+    this.tokenRef = new AtomicReference<>(null);
+  }
+
+  @VisibleForTesting
+  CrtAuthProvider(final RequestDispatcher dispatcher,
+                  final String authServer,
+                  final String username,
+                  final Signer signer) {
+    this.dispatcher = dispatcher;
+    this.username = username;
+    this.crtClient = new CrtAuthClient(signer, authServer);
+    this.tokenRef = new AtomicReference<>(null);
+  }
+
+  @Override
+  public String currentAuthorizationHeader() {
+    return tokenRef.get();
+  }
+
+  @Override
+  public ListenableFuture<String> renewAuthorizationHeader(final String ignored) {
+    final String authRequest = CrtAuthClient.createRequest(username);
+    try {
+      final HeliosRequest request = HeliosRequest.builder()
+          .method("GET")
+          .uri(new URI(AUTH_URI))
+          .appendHeader(CRT_HEADER, "request:" + authRequest)
+          .build();
+
+      final ListenableFuture<Response> challengeFuture = dispatcher.request(request);
+
+      final ListenableFuture<Response> tokenFuture =
+          Futures.transform(challengeFuture, new AsyncFunction<Response, Response>() {
+            @Override
+            public ListenableFuture<Response> apply(@NotNull final Response response)
+                throws Exception {
+              checkStatus(response, 200);
+
+              final String challenge = getHeader(response, CRT_HEADER, CHALLENGE_PREFIX);
+              final String crtResponse = crtClient.createResponse(challenge);
+              final HeliosRequest request = HeliosRequest.builder()
+                  .method("GET")
+                  .uri(new URI(AUTH_URI))
+                  .appendHeader(CRT_HEADER, "response:" + crtResponse)
+                  .build();
+
+              return dispatcher.request(request);
+            }
+          });
+      return Futures.transform(tokenFuture, new Function<Response, String>() {
+        @Nullable
+        @Override
+        public String apply(final Response response) {
+          try {
+            checkStatus(response, 200);
+            final String t = getHeader(response, CRT_HEADER, TOKEN_PREFIX);
+            final String token = "chap:" + t;
+            tokenRef.set(token);
+            return token;
+          } catch (CrtAuthException e) {
+            throw Throwables.propagate(e);
+          }
+        }
+      });
+    } catch (URISyntaxException e) {
+      // This should never happen
+      throw Throwables.propagate(e);
```

Comment:
minor: is it possible to have this try-catch wrap just whatever line is potentially throwing it?